### PR TITLE
fix(@aws-cdk/aws-autoscaling): regex validation of cron extended expressions is not compliant with specs

### LINKED ISSUE #2797: bad cron validation in autoscaling scheduled actions
**Describe the bug**

In [aws-autoscaling/lib/scheduled-action.ts](https://github.com/awslabs/aws-cdk/blob/master/packages/%40aws-cdk/aws-autoscaling/lib/scheduled-action.ts#L78), the constant that validates cron expressions does not allow commas or dashes:

```
const CRON_PART = '(\\*|\\?|[0-9]+)';

const CRON_EXPRESSION = new RegExp('^' + [CRON_PART, CRON_PART, CRON_PART, CRON_PART, CRON_PART].join('\\s+') + '$');
```

This effectively makes it impossible to create certain types of ASG schedules, for example those that do something every weekday.

**To Reproduce**

Add a schedule to an AutoScalingGroup e.g. `asg.scaleOnSchedule('end-workday', {schedule: '45 12 * * 2-6', desiredCapacity: 0});`  

This will throw an error: `Input to ScheduledAction should be a cron expression, got: 45 12 * * 2-6`. 

**Expected behavior**

Cron expressions with commas and dashes do not throw errors, but are passed into cloudformation as valid data.

**Version:**

 - MacOS
 - Typescript
 - 0.33.0

### PR DESCRIPTION
**Pull request by made2591**
**Bug**

In [aws-autoscaling/lib/scheduled-action.ts](https://github.com/awslabs/aws-cdk/blob/master/packages/%40aws-cdk/aws-autoscaling/lib/scheduled-action.ts#L78), the constant that validates cron expressions is not compliant with [specs](https://docs.aws.amazon.com/AmazonCloudWatch/latest/events/ScheduledEvents.html#CronExpressions)

const CRON_PART = '(\\*|\\?|[0-9]+)';

const CRON_EXPRESSION = new RegExp('^' + [CRON_PART, CRON_PART, CRON_PART, CRON_PART, CRON_PART].join('\\s+') + '$');

This effectively makes it impossible to create certain types of ASG schedules, for example, those that do something every weekday.

To Reproduce

Add a schedule to an AutoScalingGroup e.g. asg.scaleOnSchedule('end-workday', {schedule: '45 12 * * 2-6', desiredCapacity: 0});

This will throw an error: Input to ScheduledAction should be a cron expression, got: 45 12 * * 2-6.

**Fix**

I fixed the bug by introducing more robust regex. I had to change tests accordingly: I think once the regex will be validated can be ported somewhere else and exposed to validate cron expression in all packages.

close #2797 (partially #1181)

### Pull Request Checklist

* [x] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [x] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
  - __Design__: For significant features, design document added to `design` folder
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** and module name in parens, which will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [x] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/@aws-cdk/aws-autoscaling/lib/scheduled-action.ts [hunk 1]
```diff
@@ -75,9 +75,27 @@ export interface ScheduledActionProps extends BasicScheduledActionProps {
...
+const CRON_DAY_OF_WEEK = `(\\*|\\?|L|[1-7]\\#[1-9]|
+([1-7]|MON|TUE|WED|THU|FRI|SAT|SUN)|(([1-7]|MON|TUE|WED|THU|FRI|SAT|SUN)
+(\\,|\\-)([1-7]|MON|TUE|WED|THU|FRI|SAT|SUN)))`;
+const CRON_YEAR = `(\\*|(19[78][0-9]|199[0-9]|20[0-9]{2}|21[0-8][0-9]|219[0-9])|
+(((19[78][0-9]|199[0-9]|20[0-9]{2}|21[0-8][0-9]|219[0-9]))(\\,|\\/|\\-)
+((19[78][0-9]|199[0-9]|20[0-9]{2}|21[0-8][0-9]|219[0-9]))))`;
+
+const CRON_EXPRESSION = new RegExp('^' +
+  [
+    CRON_MINUTES, CRON_HOURS, CRON_DAY_OF_MONTH, CRON_MONTH, CRON_DAY_OF_WEEK, CRON_YEAR
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-autoscaling/lib/scheduled-action.ts]: Not sure where the YEAR field is coming from. This page says:

https://docs.aws.amazon.com/autoscaling/ec2/userguide/schedule_time.html#create-sch-actions-aws-cli

> You can specify a recurrence schedule, in UTC, using the Unix cron syntax format. This format consists of five fields separated by white spaces: [Minute] [Hour] [Day_of_Month] [Month_of_Year] [Day_of_Week]

This page also does not contain it: http://crontab.org/

Did you actually run the integ tests, or did you just update the expectations manually?

Comment (made2591) [reply packages/@aws-cdk/aws-autoscaling/lib/scheduled-action.ts]: I run them: I know the year is not present in cron, but I supposed from [here](https://docs.aws.amazon.com/autoscaling/application/userguide/application-auto-scaling-scheduled-scaling.html) that *extended cron* - the one used for rules - was supported. I think I looked at wrong doc page 😕If so, then we have to move to standard cron...

Comment (rix0rrr) [reply packages/@aws-cdk/aws-autoscaling/lib/scheduled-action.ts]: Yeah. Application AutoScaling and AutoScaling are not the same.

Comment (rix0rrr) [reply packages/@aws-cdk/aws-autoscaling/lib/scheduled-action.ts]: BTW, when you're changing it, can you explicitly add in there that although YEAR  is supported for appscaling, it's not supported for instance scaling?

Comment (made2591) [reply packages/@aws-cdk/aws-autoscaling/lib/scheduled-action.ts]: Ok, I will :) Thank you for your time!

**Comment by made2591** [comment]
I’m sorry, I just noticed I removed year but I left all the other regex as aws-cron 🤣 my bad, I will have a look at cron spec to build compliant regex asap 😅

**Comment by rix0rrr** [comment]
We recently changed the way how cron rates are expressed. I think this PR is no longer necessary to express what you want to do, so I will be closing it.

The new cron passing mechanism does not have validation yet, so if you would like to adapt this PR to validate the values passed into the new mechanism, that would be appreciated!

**Comment by made2591** [comment]
Yes, I didn’t forget about this...actually I had a regex ready to validate everything, but there was still one case not correctly validated (and, I didn't have time to check the completeness tbh). I also thought, in the beginning, to avoid validation through a regex and use instead a set of regex + some logic (inspired by some code I found online) to let the whole validation process being (even only) more readable... I arrived at this:

```
^\s*($|#|\w+\s*=|(\*|(?:[0-5]?\d)(?:(?:-|/|,)(?:[0-5]?\d))?(?:,(?:[0-5]?\d)(?:(?:-|/|,)(?:[0-5]?\d))?)*)\s+(\*|(?:[01]?\d|2[0-3])(?:(?:-|/|,)(?:[01]?\d|2[0-3]))?(?:,(?:[01]?\d|2[0-3])(?:(?:-|/|,)(?:[01]?\d|2[0-3]))?)*)\s+(\*|(?:0?[1-9]|[12]\d|3[01])(?:(?:-|/|,)(?:0?[1-9]|[12]\d|3[01]))?(?:,(?:0?[1-9]|[12]\d|3[01])(?:(?:-|/|,)(?:0?[1-9]|[12]\d|3[01]))?)*)\s+(\*|(?:[1-9]|1[012])(?:(?:-|/|,)(?:[1-9]|1[012]))?(?:,(?:[1-9]|1[012])(?:(?:-|/|,)(?:[1-9]|1[012]))?)*|\*|(?:JAN|FEB|MAR|APR|MAY|JUN|JUL|AUG|SEP|OCT|NOV|DEC)(?:(?:-)(?:JAN|FEB|MAR|APR|MAY|JUN|JUL|AUG|SEP|OCT|NOV|DEC))?(?:,(?:JAN|FEB|MAR|APR|MAY|JUN|JUL|AUG|SEP|OCT|NOV|DEC)(?:(?:-)(?:JAN|FEB|MAR|APR|MAY|JUN|JUL|AUG|SEP|OCT|NOV|DEC))?)*)\s+(\*|(?:[0-7])(?:(?:-|/|,)(?:[0-7]))?(?:,(?:[0-7])(?:(?:-|/|,)(?:[0-7]))?)*|\*|(?:MON|TUE|WED|THU|FRI|SAT|SUN)(?:(?:-)(?:MON|TUE|WED|THU|FRI|SAT|SUN))?(?:,(?:MON|TUE|WED|THU|FRI|SAT|SUN)(?:(?:-)(?:MON|TUE|WED|THU|FRI|SAT|SUN))?)*)(|\s)+)$
```

that validates 

```
"* * * * *"
"8 * * * *"
"1,8 * * * *"
"1-8 * * * *"
"1/8 * * * *"
"1-3,7-9 * * * *"
"23 */2 * * *"
"23 * 0-23/2 * *"
"1,3,5,7,9 * * * *"
"0 12 * * *"
"15 10 * * *"
"15 10 * * *"
"15 10 * * *"
"15 10 * * *"
"* 14 * * *"
"0/5 14 * * *"
"0/5 14,17 * * *"
"0-5 14 * * *"
"1,44 14 * 3 WED"
"15 10 * * Mon-Fri"
"15 6 15 * *"
"15 2 * * 3"
"15 22 * * 4"
"15 3 * * 5"
"0 12 1/5 * *"
"1 11 1 11 *"
```

Excluding ```"23 * 0-23/2 * *"```.  But yes, I will have a look at the new setup... I think I will adapt the regex - it's pretty weird I didn't find something ready to validate CRON 🤣Of course if someone else arrives first is welcome ^^ ! Thank you @rix0rrr

**Pull request closed without merging**